lib/spack/spack/traverse.py

```diff
@@ -0,0 +1,414 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from collections import defaultdict
+
+import spack.spec
+
+# Export only the high-level API.
+__all__ = ["traverse_edges", "traverse_nodes", "traverse_tree"]
+
+
+class EdgeAndDepth(object):
+    def __init__(self, edge, depth):
+        self.edge = edge
+        self.depth = depth
+
+
+def sort_edges(edges):
+    edges.sort(key=lambda edge: edge.spec.name)
+    return edges
```

Comment:
I'm not entirely sure if I follow, since the PR you linked actually _introduced_ the following edge sorting code:

```python
            # This code determines direction and yields the children/parents
            if direction == 'children':
                edges = self.edges_to_dependencies
                key_fn = lambda dspec: dspec.spec.name
                succ = lambda dspec: dspec.spec
            elif direction == 'parents':
                edges = self.edges_from_dependents
                key_fn = lambda dspec: dspec.parent.name
                succ = lambda dspec: dspec.parent
            else:
                raise ValueError('Invalid traversal direction: %s' % direction)

            for dspec in sorted(edges(), key=key_fn):
```

i.e. sort edges by edge target package name.

Reply:
Ok we discussed this in the meeting today, and we decided this does no harm *for now*.  We'll need to revisit this sort for determinism when we do compilers as deps not so long from now.